prpl/attachments.c

```diff
@@ -192,3 +194,432 @@ void download_attachment(ChimeConnection *cxn, ChimeAttachment *att, AttachmentC
+	ChimeConnection *self,
+	const gchar *messaging_url,
+	PurpleXfer *xfer
+)
+{
+	AttachmentUpload *data = (AttachmentUpload*)xfer->data;
+
+	JsonBuilder *jb = json_builder_new();
+	jb = json_builder_begin_object(jb);
+	jb = json_builder_set_member_name(jb, "ContentType");
+	jb = json_builder_add_string_value(jb, data->content_type);
+	jb = json_builder_end_object(jb);
+
+	SoupURI *uri = soup_uri_new_printf(messaging_url, "/uploads");
+	JsonNode *node = json_builder_get_root(jb);
+	chime_connection_queue_http_request(
+			self, node, uri, "POST", request_upload_url_callback, xfer);
+
+	json_node_unref(node);
+	g_object_unref(jb);
+}
+
+struct FileType {
+	const gchar *file_extension;
+	const gchar *mime_type;
+};
+
+struct FileType file_types[] = {
+	{".aac", "audio/aac"},
+	{".avi", "video/x-msvideo"},
+	{".doc", "application/msword"},
+	{".docx", "application/vnd.openxmlformats-officedocument.wordprocessingml.document"},
+	{".gif", "image/gif"},
+	{".htm", "text/html"},
+	{".html", "text/html"},
+	{".ics", "text/calendar"},
+	{".jpeg", "image/jpeg"},
+	{".jpg", "image/jpeg"},
+	{".mid", "audio/midi"},
+	{".midi", "audio/midi"},
+	{".mpeg", "video/mpeg"},
+	{".odp", "application/vnd.oasis.opendocument.presentation"},
+	{".ods", "application/vnd.oasis.opendocument.spreadsheet"},
+	{".odt", "application/vnd.oasis.opendocument.text"},
+	{".oga", "audio/ogg"},
+	{".ogv", "video/ogg"},
+	{".ogx", "application/ogg"},
+	{".png", "image/png"},
+	{".pdf", "application/pdf"},
+	{".ppt", "application/vnd.ms-powerpoint"},
+	{".pptx", "application/vnd.openxmlformats-officedocument.presentationml.presentation"},
+	{".rar", "application/x-rar-compressed"},
+	{".rtf", "application/rtf"},
+	{".svg", "image/svg+xml"},
+	{".tar", "application/x-tar"},
+	{".tif", "image/tiff"},
+	{".tiff", "image/tiff"},
+	{".wav", "audio/x-wav"},
+	{".weba", "audio/webm"},
+	{".webm", "video/webm"},
+	{".webp", "image/webp"},
+	{".xhtml", "application/xhtml+xml"},
+	{".xls", "application/vnd.ms-excel"},
+	{".xlsx", "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"},
+	{".xml", "application/xml"},
+	{".zip", "application/zip"},
+	{".7z", "application/x-7z-compressed"},
+};
+
+static void get_mime_type(gchar *filename, gchar **mime_type) {
+	gchar *file_extension = g_strrstr(basename(filename), ".");
+	const gchar *content_type = NULL;
+	if (file_extension) {
+		purple_debug_misc("chime", "File extension: %s\n", file_extension);
+		for (int i = 0; i < sizeof(file_types) / sizeof(struct FileType); i++) {
+			if (!g_strcmp0(file_extension, file_types[i].file_extension)) {
+				content_type = file_types[i].mime_type;
+				break;
+			}
+		}
+	}
+	else {
+		purple_debug_misc("chime", "File has no extension\n");
+	}
+	if (!content_type) {
+		content_type = "application/unknown";
+	}
+	purple_debug_misc("chime", "Content type: %s\n", content_type);
+	*mime_type = g_strdup(content_type);
+}
+
+// TODO: This struct is duplicated on conversations.c
```

Comment:
Any suggestion? Should I move it to a common file included by both `conversations.c` and `attachments.c`?